File: NozomiAddon/26.1.x/src/main/java/net/otsutsukimiho/nozomiaddon/features/DungeonRunOverview.java
package net.otsutsukimiho.nozomiaddon.features;

import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents;

import net.minecraft.client.DeltaTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;

import net.otsutsukimiho.nozomiaddon.gui.*;
import net.otsutsukimiho.nozomiaddon.utils.*;
import static net.otsutsukimiho.nozomiaddon.NozomiAddonClient.DUNGEON;

import java.util.HashMap;
import java.util.Map;

public class DungeonRunOverview extends DraggableHudElement implements FeatureManager.Feature {
    public DungeonRunOverview() {
        super("DungeonRunOverview", 10, 10, 200, 60, 10);
    }

    public void setEnabled(boolean enabled) {
        DungeonRunOverview.enabled = enabled;
    }

    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public ItemStack getIcon() {
        return new ItemStack(Items.MAP);
    }

    private static volatile boolean enabled = false;
    public static String dungeonFloor = "§c???";
    public static String bloodOpened = "?";
    public static float portalTime = 0;
    public static long bossEntry = 0;
    public static float bloodTick = 0;
    public static boolean bloodStart = false;
    public static float ticksElapsed = 0;
    public static long dungeonStep = 0;
    public static float realElapsed = 0;
    public static float loss = 0;
    public boolean getDungeon = false;
    public static String zone = "Total";
    public static final Map<String, Float> ticks = new HashMap<>();

    public int countTick = 0;

    public static final String[] WATCHER_DIALOGUE = {
            "Things feel a little more roomy now, eh?",
            "Oh.. hello?",
            "I'm starting to get tired of seeing you around here...",
            "You've managed to scratch and claw your way here, eh?",
            "So you made it this far... interesting.",
            "Ah, we meet again...",
            "Ah, you've finally arrived."
    };

    public void createSplit() {
        ticks.clear();
        ticks.put("Total", 0F);
        ticks.put("Maxor", 0F);
        ticks.put("Storm", 0F);
        ticks.put("Terminals", 0F);
        ticks.put("Goldor", 0F);
        ticks.put("Necron", 0F);
        ticks.put("Dragons", 0F);
        ticks.put("End", 0F);
        ticks.put("Boss", 0F);
    }

    public void initClient() {
        ClientPlayConnectionEvents.JOIN.register((handler, sender, c) -> {
            dungeonFloor = "§c???";
            bloodOpened = "?";
            zone = "Total";
            portalTime = 0;
            bossEntry = 0;
            bloodTick = 0;
            bloodStart = false;
            ticksElapsed = 0;
            dungeonStep = 0;
            getDungeon = false;
            realElapsed = 0;
            loss = 0;
            createSplit();
        });

        ClientTickEvents.END_LEVEL_TICK.register(client -> {
            if (!DUNGEON.inDungeon) return;
            if (!enabled) return;
            if (!getDungeon) {
                String getFloor = DungeonUtils.getCurrentDungeon();
                countTick++;
                if (countTick >= 20) {
                    countTick = 0;
                }
                if (getFloor == null) return;
                dungeonFloor = DungeonUtils.getCurrentDungeon();
                getDungeon = true;
            } else {
                if (dungeonFloor.matches("M\\d+")) {
                    dungeonFloor = "§4§l" + dungeonFloor;
                } else {
                    dungeonFloor = "§a" + dungeonFloor;
                }
            }

            if (DUNGEON.runStarted != -1 && DUNGEON.runEnded == -1) {
                realElapsed = (float) (System.currentTimeMillis() - DUNGEON.runStarted) / 1000;
                loss = ((realElapsed - ticksElapsed / 20) * -1);
            }

        });

        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
            if (!enabled) return;
            if (!DUNGEON.inDungeon) return;
            Minecraft client = Minecraft.getInstance();
            if (client == null) return;
            if (client.player == null) return;

            String msg = message.getString();
            if (msg.startsWith("[BOSS] The Watcher: ")) {
                for (String line : WATCHER_DIALOGUE) {
                    if (msg.contains(line)) {
                        if (!bloodStart) bloodStart = true;
                        break;
                    }
                }
            }
            if (msg.contains("[BOSS] The Watcher: That will be enough for now.")) {
                client.gui.getChat().addClientSystemMessage(Component.literal("§d§lNA §f§l» §cBlood Ready!"));
            }
            if (msg.contains("[BOSS] Maxor: WELL! WELL! WELL! LOOK WHO'S HERE!")) {
                zone = "Maxor";
            }
            if (msg.contains("[BOSS] Storm: Pathetic Maxor, just like expected.")) {
                if (dungeonStep == 3) {
                    dungeonStep = 4;
                    client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §aMaxor Done: %.2fs", (ticks.getOrDefault("Maxor", 0F) / 20.0))));
                }
                zone = "Storm";
            }
            if (msg.contains("[BOSS] Goldor: Who dares trespass into my domain?")) {
                if (dungeonStep == 4) {
                    dungeonStep = 5;
                    client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §bStorm Done: %.2fs", (ticks.getOrDefault("Storm", 0F) / 20.0))));
                }
                zone = "Terminals";
            }
            if (msg.contains("The Core entrance is opening!")) {
                if (dungeonStep == 5) {
                    dungeonStep = 6;
                    client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §eTerminals Done: %.2fs", (ticks.getOrDefault("Terminals", 0F) / 20.0))));
                }
                zone = "Goldor";
            }
            if (msg.contains("[BOSS] Necron: You went further than any human before, congratulations.")) {
                if (dungeonStep == 6) {
                    dungeonStep = 7;
                    client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §7Goldor Done: %.2fs", (ticks.getOrDefault("Goldor", 0F) / 20.0))));
                }
                zone = "Necron";
            }
            if (msg.contains("[BOSS] Necron: All this, for nothing...")) {
                if (dungeonStep == 7) {
                    dungeonStep = 8;
                    client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §cNecron Done: %.2fs", (ticks.getOrDefault("Necron", 0F) / 20.0))));
                }
                zone = "Dragons";
            }
        });
    }

    public static void onPacketReceived() {
        if (!DUNGEON.inDungeon) return;
        if (!enabled) return;
        if (DUNGEON.runStarted != -1 && DUNGEON.runEnded == -1) {
            tick();
        }
    }

    public static void tick () {
        if (!DUNGEON.inDungeon) return;
        Minecraft client = Minecraft.getInstance();
        if (client == null) return;
        if (client.player == null) return;

        float tps = 1f;
        ticksElapsed += tps;
        ticks.put(zone, ticks.getOrDefault(zone, 0F) + tps);
        if (DUNGEON.bloodOpened == -1) {
            bloodOpened = getTime(System.currentTimeMillis() - DUNGEON.runStarted);
        } else {
            bloodOpened = getTime(DUNGEON.bloodOpened - DUNGEON.runStarted);
        }
        if (bloodStart) {
            bloodTick += tps;
            if (dungeonStep == 0) {
                dungeonStep = 1;
                client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §c§lBLOOD DOOR §chas been opened! §f(§a%s§f)", bloodOpened)));
            }
        }
        if (DUNGEON.watcherCleared != -1 && DUNGEON.bossEntry == -1) {
            bloodStart = false;
            portalTime += tps;
            if (dungeonStep == 1) {
                dungeonStep = 2;
                client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §cBlood Done! §f(took §a%.2fs§f)", (bloodTick / 20.0))));
            }
        }
        if (DUNGEON.bossEntry != -1) {
            bossEntry = (DUNGEON.bossEntry - DUNGEON.runStarted);
            if (dungeonStep == 2) {
                dungeonStep = 3;
                client.gui.getChat().addClientSystemMessage(Component.literal(String.format("§d§lNA §f§l» §aBoss Entry: §a%s", getTime(bossEntry))));
                if (!dungeonFloor.contains("7")) zone = "Boss";
            }
        }
    }

    @Override
    public void render(GuiGraphicsExtractor ctx, DeltaTracker tickCounter) {
        if (!enabled && !EditHudScreen.isEditMode()) return;

        float scale = this.size / 10.0f;
        String text;
        int y = 0;

        var matrices = ctx.pose();
        matrices.pushMatrix();
        ctx.pose().translate(this.x, this.y);
        ctx.pose().scale(scale, scale);

        if (!DUNGEON.inDungeon) {
            if (EditHudScreen.isEditMode()) {
                text = String.format("§e§lRun Overview§r %s §7(§a?§7)", "§c???");
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            }
        } else {
            if (DUNGEON.runStarted == -1) {
                text = String.format("§e§lRun Overview§r %s §7(§a?§7)", dungeonFloor);
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            } else {
                text = String.format("§e§lRun Overview§r %s §7(§a%s§7) §7(§b%.1fs§7)" , dungeonFloor, getTime((long) ((ticksElapsed / 20) * 1000)), loss);
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                y += 12;
                text = String.format(" §8%s Doors §7| §4BR %s", DUNGEON.openedWitherDoors, bloodOpened);
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            }
            if (dungeonStep >= 1) {
                y += 12;
                text = String.format(" §cWatcher Clear: %.2fs", (bloodTick / 20.0));
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            }
            if (dungeonStep >= 2) {
                y += 12;
                text = String.format(" §dPortal: %.2fs", (portalTime / 20.0));
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            }
            if (dungeonStep >= 3) {
                y += 12;
                text = String.format(" §aBoss Entry: %s", getTime(bossEntry));
                ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
            }
            if (dungeonFloor.contains("M7")) {
                if (dungeonStep >= 3) {
                    y += 12;
                    text = String.format(" §aMaxor: %.2fs", (ticks.getOrDefault("Maxor", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
                if (dungeonStep >= 4) {
                    y += 12;
                    text = String.format(" §bStorm: %.2fs", (ticks.getOrDefault("Storm", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
                if (dungeonStep >= 5) {
                    y += 12;
                    text = String.format(" §eTerminals: %.2fs", (ticks.getOrDefault("Terminals", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
                if (dungeonStep >= 6) {
                    y += 12;
                    text = String.format(" §7Goldor: %.2fs", (ticks.getOrDefault("Goldor", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
                if (dungeonStep >= 7) {
                    y += 12;
                    text = String.format(" §cNecron: %.2fs", (ticks.getOrDefault("Necron", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
                if (dungeonStep >= 8) {
                    y += 12;
                    text = String.format(" §4Dragons: %.2fs", (ticks.getOrDefault("Dragons", 0F) / 20.0));
                    ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                }
            } else {
                if (dungeonFloor.contains("7")) {
                    if (dungeonStep >= 3) {
                        y += 12;
                        text = String.format(" §aMaxor: %.2fs", (ticks.getOrDefault("Maxor", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                    if (dungeonStep >= 4) {
                        y += 12;
                        text = String.format(" §bStorm: %.2fs", (ticks.getOrDefault("Storm", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                    if (dungeonStep >= 5) {
                        y += 12;
                        text = String.format(" §eTerminals: %.2fs", (ticks.getOrDefault("Terminals", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                    if (dungeonStep >= 6) {
                        y += 12;
                        text = String.format(" §7Goldor: %.2fs", (ticks.getOrDefault("Goldor", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                    if (dungeonStep >= 7) {
                        y += 12;
                        text = String.format(" §cNecron: %.2fs", (ticks.getOrDefault("Necron", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                } else {
                    if (dungeonStep >= 3) {
                        y += 12;
                        text = String.format(" §4Boss: %.2fs", (ticks.getOrDefault("Boss", 0F) / 20.0));
                        ctx.text(Minecraft.getInstance().font, Component.literal(text), 0, y, 0xFFFFFFFF, true);
                    }
                }
            }
        }

        ctx.pose().popMatrix();
        this.width = Math.round(200 * scale);
        this.height = Math.round(60 * scale);
    }

    public static String getTime(Long ms) {
        if (ms == null || ms == 0) return "?";
        long minutes = ms / 60000;
        long seconds = (ms / 1000) % 60;
        if (minutes != 0) {
            return minutes + "m " + seconds + "s";
        } else {
            return seconds + "s";
        }
    }

}

File: NozomiAddonCT/features/DungeonRunOverview.js
import Settings from "../config";
import Dungeon from "../../BloomCore/dungeons/Dungeon";
import { getBloodTicks } from "./DungeonBlood";
import { onChatPacket } from "../../BloomCore/utils/Events";
import { registerHudElement } from "../utils/HUDManager";
import { getTime, getSecs, registerWhen } from "../utils/utils";

const S32PacketConfirmTransaction = Java.type("net.minecraft.network.play.server.S32PacketConfirmTransaction")

let mainText = new Text(` `, 5, 5).setScale(1).setShadow(true);

let overviewStr = " "
let ticksElapsed = 0;
let splittimer = {
    Step: 0,
    Zone: "Total",
    Total: 0,
    Maxor: 0,
    Storm: 0,
    Terminals: 0,
    Goldor: 0,
    Necron: 0,
    Dragons: 0
};

let isDragging = false;
let dragOffsetX = 0;
let dragOffsetY = 0;

registerWhen(register("packetReceived", () => {
    if (!Settings.runoverview) return
    if (!Dungeon.inDungeon) return;

    let bloodOpened = "?"
    let portalTime = "?"
    let bossEntry = "?"

    if (!Dungeon.bloodOpened && Dungeon.runStarted) bloodOpened = getTime(Date.now() - Dungeon.runStarted)
    else if (Dungeon.bloodOpened) bloodOpened = getTime(Dungeon.bloodOpened - Dungeon.runStarted)
    if (!Dungeon.bloodOpened && Dungeon.runStarted && splittimer.Step == 0) { } else if (Dungeon.bloodOpened && splittimer.Step == 0) { splittimer.Step = 1 };

    if (Dungeon.watcherCleared && !Dungeon.bossEntry) portalTime = `${Math.floor((Date.now() - Dungeon.watcherCleared) / 10) / 100}s`
    if (Dungeon.bossEntry) portalTime = `${Math.floor((Dungeon.bossEntry - Dungeon.watcherCleared) / 10) / 100}s`
    if (Dungeon.watcherCleared && !Dungeon.bossEntry && splittimer.Step == 1) { splittimer.Step = 2 };

    if (Dungeon.bossEntry || !Dungeon.runStarted) bossEntry = getTime(Dungeon.bossEntry - Dungeon.runStarted)
    if (Dungeon.bossEntry && splittimer.Step == 2 || !Dungeon.runStarted && splittimer.Step == 2) { splittimer.Step = 3 };

    let dungeonFloor = "???"
    let dungeonText = "???"
    if (Dungeon.dungeonType == "Master Mode") {
        dungeonFloor = `&4M${Dungeon.floorNumber}`;
        dungeonText = `M${Dungeon.floorNumber}`;
    } else {
        if (Dungeon.floorNumber == 0) { dungeonFloor = "&aE" } else {
            dungeonFloor = `&aF${Dungeon.floorNumber}`;
            dungeonText = `F${Dungeon.floorNumber}`;
        };
    };

    let bloodticks = getBloodTicks()
    let realElapsed = (Date.now() - Dungeon.runStarted) / 1000;
    let tickElapsed = ticksElapsed / 20;
    let loss = ((realElapsed - tickElapsed) * -1).toFixed(1);

    if (Dungeon.runStarted && !Dungeon.runEnded) {
        ticksElapsed++;
        splittimer[splittimer.Zone]++;
        realElapsed = (Date.now() - Dungeon.runStarted) / 1000;
        tickElapsed = ticksElapsed / 20;
        loss = ((realElapsed - tickElapsed) * -1).toFixed(1);
    } else {
        realElapsed = (Dungeon.runEnded - Dungeon.runStarted) / 1000;
        tickElapsed = ticksElapsed / 20;
        loss = ((realElapsed - tickElapsed) * -1).toFixed(1);
    };

    if (!Dungeon.bossEntry && Dungeon.runStarted) {
        overviewStr = `&e&lRun Overview&r ${dungeonFloor} &7(&a${getTime((ticksElapsed / 20) * 1000)}&7) &7(&b${loss}s&7)`
        overviewStr += `\n &8${Dungeon.openedWitherDoors} Doors &7| &4BR ${bloodOpened}`
        if (splittimer.Step >= 1) overviewStr += `\n &cWatcher Clear: ${(bloodticks / 20).toFixed(2)}s`
        if (splittimer.Step >= 2) overviewStr += `\n &dPortal: ${portalTime}`
        if (splittimer.Step >= 3) overviewStr += `\n &aBoss Entry: ${bossEntry}`
    } else {
        if (Dungeon.bossEntry && Dungeon.runStarted) {
            if (dungeonText == "M7" || dungeonText == "F7") {
                overviewStr = `&e&lRun Overview&r ${dungeonFloor} &7(&a${getTime((ticksElapsed / 20) * 1000)}&7) &7(&b${loss}s&7)`
                overviewStr += `\n &8${Dungeon.openedWitherDoors} Doors &7| &4BR ${bloodOpened}`
                if (splittimer.Step >= 1) overviewStr += `\n &cWatcher Clear: ${(bloodticks / 20).toFixed(2)}s`
                if (splittimer.Step >= 2) overviewStr += `\n &dPortal: ${portalTime}`
                if (splittimer.Step >= 3) overviewStr += `\n &aBoss Entry: ${bossEntry}`
                if (splittimer.Step >= 4) overviewStr += `\n &aMaxor: ${(splittimer.Maxor / 20).toFixed(2)}s`
                if (splittimer.Step >= 5) overviewStr += `\n &bStorm: ${(splittimer.Storm / 20).toFixed(2)}s`
                if (splittimer.Step >= 6) overviewStr += `\n &eTerminals: ${(splittimer.Terminals / 20).toFixed(2)}s`
                if (splittimer.Step >= 7) overviewStr += `\n &7Goldor: ${(splittimer.Goldor / 20).toFixed(2)}s`
                if (splittimer.Step >= 8) overviewStr += `\n &cNecron: ${(splittimer.Necron / 20).toFixed(2)}s`
                if (splittimer.Step >= 9) overviewStr += `\n &4Dragons: ${(splittimer.Dragons / 20).toFixed(2)}s`
            } else {
                overviewStr = `&e&lRun Overview&r ${dungeonFloor} &7(&a${getTime((ticksElapsed / 20) * 1000)}&7) &7(&b${loss}s&7)`
                overviewStr += `\n &8${Dungeon.openedWitherDoors} Doors &7| &4BR ${bloodOpened}`
                if (splittimer.Step >= 1) overviewStr += `\n &cWatcher Clear: ${(bloodticks / 20).toFixed(2)}s`
                if (splittimer.Step >= 2) overviewStr += `\n &dPortal: ${portalTime}`
                if (splittimer.Step >= 3) overviewStr += `\n &aBoss Entry: ${bossEntry}`
            };
        } else {
            overviewStr = `&e&lRun Overview &7(${dungeonFloor}&7)`
            overviewStr += `\n &7Waiting...`
        };
    };

}).setFilteredClass(S32PacketConfirmTransaction), () => Settings.runoverview && Dungeon.inDungeon);

onChatPacket(() => { splittimer.Step = 4; splittimer.Zone = "Maxor" }).setCriteria("[BOSS] Maxor: WELL! WELL! WELL! LOOK WHO'S HERE!");
onChatPacket(() => { splittimer.Step = 5; splittimer.Zone = "Storm"; ChatLib.chat(`${Settings.prefix} &aMaxor Done: ${(splittimer.Maxor / 20).toFixed(2)}s`) }).setCriteria("[BOSS] Storm: Pathetic Maxor, just like expected.");
onChatPacket(() => { splittimer.Step = 6; splittimer.Zone = "Terminals"; ChatLib.chat(`${Settings.prefix} &bStorm Done: ${(splittimer.Storm / 20).toFixed(2)}s`) }).setCriteria("[BOSS] Goldor: Who dares trespass into my domain?");
onChatPacket(() => { splittimer.Step = 7; splittimer.Zone = "Goldor"; ChatLib.chat(`${Settings.prefix} &eTerminals Done: ${(splittimer.Terminals / 20).toFixed(2)}s`) }).setCriteria("The Core entrance is opening!");
onChatPacket(() => { splittimer.Step = 8; splittimer.Zone = "Necron"; ChatLib.chat(`${Settings.prefix} &7Goldor Done: ${(splittimer.Goldor / 20).toFixed(2)}s`) }).setCriteria("[BOSS] Necron: You went further than any human before, congratulations.");
onChatPacket(() => { splittimer.Step = 9; splittimer.Zone = "Dragons"; ChatLib.chat(`${Settings.prefix} &cNecron Done: ${(splittimer.Necron / 20).toFixed(2)}s`) }).setCriteria("[BOSS] Necron: All this, for nothing...");

registerHudElement(() => {
    if (!Settings.runoverview) return;
    let x = Number(Settings.runoverview_pos_x);
    let y = Number(Settings.runoverview_pos_y);
    let text = `${overviewStr}`
    mainText.setString(text);
    if (Settings.editUI) {
        let textWidth = mainText.getWidth();
        let textHeight = mainText.getHeight();
        if (Settings.editUIName.includes("RunOverview")) {
            Renderer.drawRect(Renderer.color(0, 255, 0, 100), x - 3, y - 3, textWidth + 6, textHeight + 6);
        } else {
            Renderer.drawRect(Renderer.color(255, 255, 255, 150), x - 3, y - 3, textWidth + 6, textHeight + 6);
        };
    };
    mainText.setScale(Settings.runoverview_scale);
    mainText.draw(x, y);
}, () => Settings.runoverview);

registerWhen(register("clicked", (x, y, button, isDown) => {
    if (!Settings.runoverview) return;
    if (!Settings.editUI) return
    if (!isDown) {
        Settings.editUIName = Settings.editUIName.replace("RunOverview", " ").trim();
        isDragging = false;
        return;
    };

    let textX = Number(Settings.runoverview_pos_x);
    let textY = Number(Settings.runoverview_pos_y);
    let textWidth = mainText.getWidth();
    let textHeight = mainText.getHeight();

    if (button === 0) {
        if (x >= textX && x <= textX + textWidth && y >= textY && y <= textY + textHeight) {
            isDragging = true;
            dragOffsetX = x - textX;
            dragOffsetY = y - textY;
            if (Settings.editUIName == "") Settings.editUIName += "RunOverview"
        };
    } else if (button === 1) {
        if (!Settings.editUIName.includes("RunOverview")) return
        World.playSound("tile.piston.in", 100, 1);
        Settings.runoverview_pos_x = "0";
        Settings.runoverview_pos_y = "0";
        Settings.runoverview_scale = "1";
        Settings.save()
    };
}), () => Settings.runoverview);

registerWhen(register("dragged", (dx, dy, x, y) => {
    if (!isDragging) return;
    if (!Settings.editUI) return
    if (!Settings.editUIName.includes("RunOverview")) return
    Settings.runoverview_pos_x = x - dragOffsetX;
    Settings.runoverview_pos_y = y - dragOffsetY;
}), () => Settings.runoverview);

registerWhen(register("scrolled", (x, y, dir) => {
    if (!Settings.editUI) return
    if (!Settings.editUIName.includes("RunOverview")) return
    Settings.runoverview_scale = Number(Settings.runoverview_scale)
    if (dir == 1) Settings.runoverview_scale += 0.05
    else Settings.runoverview_scale -= 0.05
}), () => Settings.runoverview);

registerWhen(register("worldUnload", () => {
    ticksElapsed = 0;
    overviewStr = " "
    splittimer = {
        Step: 0,
        Zone: "Total",
        Total: 0,
        Maxor: 0,
        Storm: 0,
        Terminals: 0,
        Goldor: 0,
        Necron: 0,
        Dragons: 0
    };
}), () => Settings.runoverview);

registerWhen(register("worldLoad", () => {
    ticksElapsed = 0;
    overviewStr = " "
    splittimer = {
        Step: 0,
        Zone: "Total",
        Total: 0,
        Maxor: 0,
        Storm: 0,
        Terminals: 0,
        Goldor: 0,
        Necron: 0,
        Dragons: 0
    };
}), () => Settings.runoverview);